### PR TITLE
Revert the signature for `TaskInternal.getOnlyIf` to return `Spec`

### PR DESCRIPTION
**Pull request by h0tk3y**
Fixes https://github.com/gradle/gradle/issues/23520

### CODE REVIEW

**Comment by h0tk3y** [comment]
@bot-gradle test RFN

File: build-logic/instrumentation/src/main/kotlin/gradlebuild/instrumentation/classpostprocessing/configure.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    tasks.withType(JavaCompile::class.java).named(compileJavaTaskName) {
+        configureTask(extension, AbstractCompile::getDestinationDirectory)
+    }
+}
+
+
+internal
+fun <T : Task> T.configureTask(extension: PostProcessCompiledClassesExtension, taskClassesDir: (T) -> Provider<Directory>) {
+    inputs.property("classPostProcessors", extension.classPostProcessors)
+    doLast {
```
Comment (mlopatkin) [inline build-logic/instrumentation/src/main/kotlin/gradlebuild/instrumentation/classpostprocessing/configure.kt:39]: Does this `doLast` work well with build cache? I'm not sure whether any unrelated change to the build scripts invalidates the build cache because of it.

File: build-logic/instrumentation/src/main/kotlin/gradlebuild/instrumentation/classpostprocessing/postProcessClasses.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import org.objectweb.asm.ClassWriter
+import java.io.File
+
+
+internal
+fun postProcessClasses(
+    classesDirectory: File,
+    postProcessors: Iterable<ClassPostProcessor>
+) {
+    postProcessors.forEach { postProcessor ->
```
Comment (mlopatkin) [inline build-logic/instrumentation/src/main/kotlin/gradlebuild/instrumentation/classpostprocessing/postProcessClasses.kt:30]: It would be nice to fan out with Worker API, but this doesn't matter much at this point, as we only have a couple of classes to process.

**Review by mlopatkin — APPROVED** [review]
Just a couple suggestions, but nothing particularly critical.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipOnlyIfTaskExecuter.java [hunk 3]
```diff
@@ -41,7 +43,11 @@ public SkipOnlyIfTaskExecuter(TaskExecuter executer) {
     public TaskExecuterResult execute(TaskInternal task, TaskStateInternal state, TaskExecutionContext context) {
         SelfDescribingSpec<? super TaskInternal> unsatisfiedSpec;
         try {
-            unsatisfiedSpec = task.getOnlyIf().findUnsatisfiedSpec(task);
+            Spec<? super TaskInternal> onlyIf = task.getOnlyIf();
+            if (!(onlyIf instanceof DescribingAndSpec)) {
+                throw new IllegalStateException("unexpected type of onlyIf spec: " + onlyIf.getClass());
```
[the remark this answers is not in the document]
Comment (h0tk3y) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipOnlyIfTaskExecuter.java:48]: Indeed, I think we should do this. I found some community plugins overriding `getOnlyIf` in the tasks, so we should support returning the more generic `Spec` here...

File: subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java [hunk 4]
```diff
@@ -385,7 +385,7 @@ public boolean isSatisfiedBy(Task element) {
     }
 
     @Override
-    public DescribingAndSpec<? super TaskInternal> getOnlyIf() {
+    public Spec<? super TaskInternal> getOnlyIf() {
```
[the remark this answers is not in the document]
Comment (h0tk3y) [reply subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java:388]: Keeping the `DescribingAndSpec`-returning method in the sources will lead to plugins using this API and compiled against Gradle >= 7.6 becoming incompatible with Gradle < 7.6 runtime for no good reason, do you think we can accept this?
Also, yes, there are several community plugins overriding `getOnlyIf` in the tasks.

Maybe we should get better at warning users about internal APIs leaking through `DefaultTask`, so they are aware of the stuff they should not use or override.

[the remark this answers is not in the document]
Comment (mlopatkin) [reply subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java:388]: >will lead to plugins using this API and compiled against Gradle >= 7.6 becoming incompatible with Gradle < 7.6 runtime for no good reason

Oh, I see, I didn't think through this aspect of backwards compatibility. I was mostly trying to stick to the javac-blessed pattern to avoid unforeseen problems.

**Comment by h0tk3y** [comment]
@bot-gradle test and merge
